File: backend/app/models/coach.py
"""Pydantic models for Coach Command Center API responses."""

from typing import Any
from pydantic import BaseModel


class AssignmentTemplateRow(BaseModel):
    id: str
    team_id: str | None = None
    created_by: str | None = None
    title: str
    description: str | None = None
    kind: str = "speech"
    speech_type: str | None = None
    target_skill: str | None = None
    success_criteria: list[str] = []
    goal: str | None = None
    duration_minutes: int | None = None
    due_offset_days: int = 7
    is_built_in: bool = False
    created_at: str | None = None


class AssignmentTemplateCreate(BaseModel):
    team_id: str
    title: str
    description: str | None = None
    kind: str = "speech"
    speech_type: str | None = None
    target_skill: str | None = None
    success_criteria: list[str] = []
    goal: str | None = None
    duration_minutes: int | None = None
    due_offset_days: int = 7


class AssignTemplateRequest(BaseModel):
    team_id: str
    recipient_user_ids: list[str]
    due_date: str | None = None


class AttentionFlag(BaseModel):
    rule: str
    reason: str
    link: str


class RosterRow(BaseModel):
    user_id: str
    display_name: str | None = None
    last_practice_at: str | None = None
    days_inactive: int | None = None
    speech_count: int = 0
    active_mission_skill: str | None = None
    active_mission_status: str | None = None
    active_mission_id: str | None = None
    priority_skill: str | None = None
    score_trend: str = "insufficient"
    top_assignment_status: str | None = None
    drill_attempts_count: int = 0
    drills_assigned_count: int = 0
    attention_level: str = "none"
    attention_flags: list[AttentionFlag] = []


class TeamSummary(BaseModel):
    total_students: int
    practiced_this_week: int
    practiced_this_month: int
    without_recent_session: int
    pending_reviews: int
    overdue_assignments: int
    avg_readiness_pct: int | None = None


class CommandCenterResponse(BaseModel):
    team_id: str
    team_name: str
    invite_code: str
    summary: TeamSummary
    roster: list[RosterRow]
    attention_queue: list[RosterRow]
    onboarding_done: bool = False


class WeeklyReportRosterItem(BaseModel):
    user_id: str
    display_name: str | None = None
    speeches_this_week: int = 0
    drills_this_week: int = 0
    active_mission_skill: str | None = None


class WeeklyReportResponse(BaseModel):
    period_start: str
    period_end: str
    students_participated: int
    total_students: int
    speeches_analyzed: int
    drills_completed: int
    assignments_completed: int
    students_improving: int
    students_needing_attention: int
    common_team_weakness: str | None = None
    recommended_focus: str
    roster: list[WeeklyReportRosterItem]


class UsageSummaryResponse(BaseModel):
    team_id: str
    active_seats: int
    speeches_analyzed: int
    evidence_searches: int
    drills_completed: int
    assignments_created: int
    storage_used_mb: float


class CoachNote(BaseModel):
    id: str
    team_id: str
    coach_id: str
    student_id: str
    note: str
    created_at: str


class CoachNoteCreate(BaseModel):
    note: str


class RichStudentProfile(BaseModel):
    student_id: str
    display_name: str | None = None
    speech_count: int
    feedback_ready_count: int
    speeches: list[dict[str, Any]] = []
    assignments: list[dict[str, Any]] = []
    active_mission: dict[str, Any] | None = None
    completed_missions: list[dict[str, Any]] = []
    drill_attempts_count: int = 0
    drills_assigned_count: int = 0
    coach_notes: list[CoachNote] = []
    attention_flags: list[AttentionFlag] = []
    roster_row: RosterRow | None = None

File: frontend/src/types/coach.ts
/** Coach Command Center types (Pass 20). */

export interface AttentionFlag {
  rule: string;
  reason: string;
  link: string;
}

export interface RosterRow {
  user_id: string;
  display_name: string | null;
  last_practice_at: string | null;
  days_inactive: number | null;
  speech_count: number;
  active_mission_skill: string | null;
  active_mission_status: string | null;
  active_mission_id: string | null;
  priority_skill: string | null;
  score_trend: "improving" | "declining" | "steady" | "insufficient";
  top_assignment_status: string | null;
  drill_attempts_count: number;
  drills_assigned_count: number;
  attention_level: "none" | "medium" | "high";
  attention_flags: AttentionFlag[];
}

export interface TeamSummary {
  total_students: number;
  practiced_this_week: number;
  practiced_this_month: number;
  without_recent_session: number;
  pending_reviews: number;
  overdue_assignments: number;
  avg_readiness_pct: number | null;
}

export interface CommandCenterData {
  team_id: string;
  team_name: string;
  invite_code: string;
  summary: TeamSummary;
  roster: RosterRow[];
  attention_queue: RosterRow[];
  onboarding_done: boolean;
}

export interface WeeklyReportRosterItem {
  user_id: string;
  display_name: string | null;
  speeches_this_week: number;
  drills_this_week: number;
  active_mission_skill: string | null;
}

export interface WeeklyReport {
  period_start: string;
  period_end: string;
  students_participated: number;
  total_students: number;
  speeches_analyzed: number;
  drills_completed: number;
  assignments_completed: number;
  students_improving: number;
  students_needing_attention: number;
  common_team_weakness: string | null;
  recommended_focus: string;
  roster: WeeklyReportRosterItem[];
}

export interface UsageSummary {
  team_id: string;
  active_seats: number;
  speeches_analyzed: number;
  evidence_searches: number;
  drills_completed: number;
  assignments_created: number;
  storage_used_mb: number;
}

export interface AssignmentTemplate {
  id: string;
  team_id: string | null;
  created_by: string | null;
  title: string;
  description: string | null;
  kind: "speech" | "drill" | "rerecord";
  speech_type: string | null;
  target_skill: string | null;
  success_criteria: string[];
  goal: string | null;
  duration_minutes: number | null;
  due_offset_days: number;
  is_built_in: boolean;
  created_at: string | null;
}

export interface CoachNote {
  id: string;
  team_id: string;
  coach_id: string;
  student_id: string;
  note: string;
  created_at: string;
}

export interface RichStudentProfile {
  student_id: string;
  display_name: string | null;
  speech_count: number;
  feedback_ready_count: number;
  speeches: Array<{
    id: string;
    title: string;
    speech_type: string;
    status: string;
    created_at: string;
  }>;
  assignments: Array<{
    recipient_id: string;
    title: string;
    status: string;
    submission_speech_id: string | null;
    coach_feedback: string | null;
    reviewed_at: string | null;
  }>;
  active_mission: Record<string, unknown> | null;
  completed_missions: Array<Record<string, unknown>>;
  drill_attempts_count: number;
  drills_assigned_count: number;
  coach_notes: CoachNote[];
  attention_flags: AttentionFlag[];
  roster_row: RosterRow | null;
}

export const SKILL_LABEL: Record<string, string> = {
  weighing: "Weighing",
  extensions: "Extensions",
  drops: "Drops",
  clash: "Clash",
  judge_adaptation: "Judge Adapt.",
  delivery: "Delivery",
  warranting: "Warranting",
  evidence_use: "Evidence",
  organization: "Organization",
};

export const TREND_LABEL: Record<string, string> = {
  improving: "Improving",
  declining: "Declining",
  steady: "Steady",
  insufficient: "New",
};

export const TREND_COLOR: Record<string, string> = {
  improving: "text-ok",
  declining: "text-danger",
  steady: "text-ink-subtle",
  insufficient: "text-ink-subtle",
};
